### PR TITLE
New binding to control RC switches

### PR DESCRIPTION
**Pull request by mroeckl**
This OpenHAB binding can control RC switches on a Raspberry Pi via a 433Mhz RF module connected to one of the IO ports of the Raspberry Pi.

The IO port can be configured in the OpenHAB configuration file. Items can be set up like this:

*Switch   SleepingRoomLight   {rcswitch="11111:4"}*

*'11111'* being the group address and *'4'* being the device address.

A discussion is already ongoing in the OpenHAB issues/form: [RC Switch binding to control RF transmitters](https://github.com/openhab/openhab/issues/1993) and [Binding to control RF Power Switches](https://groups.google.com/forum/#!category-topic/openhab/features/_TPI5STjP8s)

### CODE REVIEW

**Comment by teichsta** [comment]
Thanks @mroeckl for this great contribution. Since we are heavily working towards the 1.7 Release (May 25th) i won't be able to review this PR until beginning of June. Best, Thomas E.-E.

**Comment by haidermirza** [comment]
Is this binding only going to transmit via the RF module. I need the openhab on my RaspberryPi to receive 433 MHz data. is it possible with this??

**Comment by mroeckl** [comment]
Unfortunately, it is not possible with the current implementation to receive data. But, generally, it should be possible to implement that with the used library. I will give it a try when I have time.

**Comment by haidermirza** [comment]
Can you suggest any other way that i can use to send and receive data through the RF module..

**Comment by teichsta** [comment]
Hi @mroeckl, sorry for letting you waiting so long for the review. Will start now and let you know i am finished. Best, Thomas E.-E.

File: bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,32 @@
+Manifest-Version: 1.0
+Private-Package: org.openhab.binding.rcswitch.internal
+Ignore-Package: org.openhab.binding.rcswitch.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB RC Switch Binding
+Bundle-SymbolicName: org.openhab.binding.rcswitch
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.1.0.qualifier
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF]: bump version to 1.8

File: bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,32 @@
+Manifest-Version: 1.0
+Private-Package: org.openhab.binding.rcswitch.internal
+Ignore-Package: org.openhab.binding.rcswitch.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB RC Switch Binding
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF]: rename binding to something more specific like "rpircswitch” "openHAB Raspberry Pi RC Switch Binding"?

File: bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+Bundle-SymbolicName: org.openhab.binding.rcswitch
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.1.0.qualifier
+Bundle-Activator: org.openhab.binding.rcswitch.internal.RcSwitchActivator
+Bundle-ManifestVersion: 2
+Bundle-Description: This is the RC Switch binding of the open Home Aut
+ omation Bus (openHAB)
+Import-Package:  org.apache.commons.codec.binary;version="1.3.0",
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF]: remove version if not explicitly needed

File: bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ org.osgi.framework,
+ org.osgi.service.cm,
+ org.osgi.service.component,
+ org.osgi.service.event,
+ org.slf4j
+Export-Package: org.openhab.binding.rcswitch
+Bundle-DocURL: http://www.openhab.org
+Service-Component: OSGI-INF/binding.xml, OSGI-INF/genericbindingprovider.xml
+Bundle-ClassPath: .,
+ lib/pi4j-core.jar
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF]: any chance to specify the version of the lib?

File: bundles/binding/org.openhab.binding.rcswitch/pom.xml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.openhab.bundles</groupId>
+		<artifactId>binding</artifactId>
+		<version>1.7.0-SNAPSHOT</version>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/pom.xml]: bump version to 1.8

File: bundles/binding/org.openhab.binding.rcswitch/pom.xml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	</parent>
+
+	<name>openHAB RC Switch Binding</name>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/pom.xml]: see above - change name

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/de/pi3g/pi/rcswitch/RCSwitch.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301  USA
+ */
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/de/pi3g/pi/rcswitch/RCSwitch.java]: using GPL licensed code with EPL licensed openHAB doesn't fit very well (see https://eclipse.org/legal/eplfaq.php#GPLCOMPATIBLE). You would either have to relicense the while binding to GPL or find another way with this (single) class.

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/RcSwitchBindingProvider.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.openhab.binding.rcswitch.internal.RcSwitchBindingConfig;
+import org.openhab.core.binding.BindingProvider;
+
+/**
+ * This interface is implemented by classes that can provide mapping information
+ * between openHAB items and RcSwitch devices.
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/RcSwitchBindingProvider.java]: 1.8

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchActivator.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import org.osgi.framework.BundleActivator;
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Extension of the default OSGi bundle activator
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchActivator.java]: 1.8

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+import de.pi3g.pi.rcswitch.RCSwitch;
+
+/**
+ * <p>
+ * Binding listening OpenHAB bus and send commands to RC switches when command
+ * is received.
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java]: 1.8

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+						this.transmitter = new RCSwitch(gpioPin);
+					} catch (NumberFormatException e) {
+						LOGGER.error("Invalid configuration. Please provide an Integer value in the 'gpioPin' configuration");
+					}
+				}
+			}
+		}
+	}
+
+}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java]: add newline

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+import com.pi4j.io.gpio.PinPullResistance;
+import com.pi4j.io.gpio.RaspiGpioProvider;
+import com.pi4j.io.gpio.impl.PinImpl;
+
+import de.pi3g.pi.rcswitch.RCSwitch;
+
+/**
+ * <p>
+ * Binding listening OpenHAB bus and send commands to RC switches when command
+ * is received.
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java]: please also provide some examples of valid (and invalid) *.items configurations

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBindingConfig.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.BitSet;
+
+import org.openhab.core.binding.BindingConfig;
+
+/**
+ * An {@link RcSwitchBindingConfig} stores group and device address of one RC
+ * switch.
+ * 
+ * @author Matthias Röckl
+ * @since 1.1.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBindingConfig.java]: 1.8

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchGenericBindingProvider.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+ * 
+ * <p>
+ * Examples for valid binding configuration strings:
+ * 
+ * <ul>
+ * <li><code>rcswitch="10101:4"</code></li>
+ * </ul>
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchGenericBindingProvider.java]: 1.8

**Comment by teichsta** [comment]
your binding looks quite good to me already. Please find some review comments inline. Also:

* add this binding to ```distribution/pom.xml```
* add this binding to ```bundles/binding/pom.xml```
* add a wiki page to describe the purpose of this binding

Thanks again for your contribution @mroeckl please keep in mind to squash your commits into one once you've incorporated my feedback.

Best, Thomas E.-E.

File: bundles/binding/org.openhab.binding.rcswitch/src/main/java/de/pi3g/pi/rcswitch/RCSwitch.java (hunk 7 above)
Comment (kaikreuzer) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/de/pi3g/pi/rcswitch/RCSwitch.java]: > You would either have to relicense the while binding to GPL

@teichsta : Please note that this option means that the binding cannot be included to the addons.zip, but would have to be distributed separately, so this isn't really an option.

Comment (teichsta) [inline bundles/binding/org.openhab.binding.rcswitch/src/main/java/de/pi3g/pi/rcswitch/RCSwitch.java]: right, no good option. But it is an option in favour of "loosing" this work

**Comment by steve-bate** [comment]
@mroeckl Any updates on this PR?

**Comment by mroeckl** [comment]
@steve-bate I am currently working on the license issue. Hopefully I can solve it quickly.

**Comment by teichsta** [comment]
> I am currently working on the license issue

so, what is your plan here, @mroeckl?

**Comment by mroeckl** [comment]
I am in discussion with the original author of this class to check if he is willing to adapt the license

**Comment by teichsta** [comment]
great! looking forward to seeing this solved … if you won't succeed i would assume the best thing would be to reimplement the class (it's "just" one if i remember correctly?)

**Comment by dominicdesu** [comment]
@teichsta @mroeckl 
The included pi4j library is as well LGPL (http://pi4j.com/license.html).
As mentioned in the news (http://pi4j.com/index.html), pi4j license was recently changed from Apache to LGPL because the embedded wiringPi library is using LGPL.

Is there a difference between directly included source code and included jar libraries regarding the licenses?

**Comment by JensErat** [comment]
Using an LGPL library means you can link against the unmodified library while using any arbitrary other license for your main product. If you modify the library, you have to provide those changes under LGPL.

LGPL libraries in OpenHAB does not seem to be an issue, considering [the Google TTS bundle using an LGPL library was merged](https://github.com/openhab/openhab/pull/1950).

**Comment by kaikreuzer** [comment]
> LGPL libraries in OpenHAB does not seem to be an issue, considering the Google TTS bundle using an LGPL library was merged.

Or it does mean that the reviewer didn't pay much attention to it. In general, openHAB's EPL license is not really compatible with the LGPL, it is e.g. not allowed for Eclipse projects see https://mmilinkov.wordpress.com/2009/04/30/lgpl-pain/.
As openHAB is not a commercial system, I tend to not see it as strongly as Mike. But it must be clear that including LGPL code is an issue for anybody who uses openHAB commercially (which I never recommended to anybody). My suggestion would be to still clearly state that the bundle contains 3rd party LGPL code by adding an about.html (like https://github.com/openhab/openhab2/blob/master/addons/binding/org.openhab.binding.freebox/about.html)

**Comment by mroeckl** [comment]
@teichsta 
I committed the following changes within a single commit:

- increased version to 1.8.0
- renamed symbolic name to org.openhab.binding.rpircswitch
- renamed the binding name to 'openHAB Raspberry Pi RC Switch Binding'
- renamed packages to rpircswitch instead rcswitch
- renamed classed to RPiRcSwitch* instead of RcSwitch*
- removed version of Import-Package
- added version number to Pi4J library
- added binding to distribution/pom.xml
- added binding to bundles/binding/pom.xml
- added an about.html with license information

Let me know if shall change more than that.

One more question: Shall I add a wiki page directly or wait until the binding is released?

Thank you for your support.

**Comment by teichsta** [comment]
> Let me know if shall change more than that.

yes, could you please resolve the current merge conflict. And please squash your commits into after all.

> One more question: Shall I add a wiki page directly or wait until the binding is released?

Yes, please add the wiki page directly with a small note that this binding will be available in the official distribution with the upcoming 1.8 Release.

Thanks, Thomas E.-E.

**Comment by mroeckl** [comment]
@teichsta I merged the upstream and solved the conflicts. Now, I am trying to squash the commits. Not sure if this is gonna work to squash my last 5 commits:
-# git reset --soft HEAD~5 && git commit
Can I do the squash like that?

**Comment by watou** [comment]
[This](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) is a good description I refer to for rebasing and squashing commits.  Some labels have to be changed to apply to openHAB (like `edx` would be `upstream`), but it's a good step-by-step.

**Comment by mroeckl** [comment]
Thanks @watou . I started from scratch following your link. Hope I did everything correct now. If not, please let me know.

**Comment by teichsta** [comment]
btw:

> Or it does mean that the reviewer didn't pay much attention to it.

sorry for that! And thanks @dominicdesu for fixing this license issue by adding the about page!

**Comment by teichsta** [comment]
Thanks to all helping this PR done!

**Comment by teichsta** [comment]
@mroeckl after all could you please provide the according Wiki page until the weekend? Thanks, Thomas E.-E.

**Comment by mroeckl** [comment]
@teichsta I added a wiki page called 'Raspberry Pi RC Switch Binding' and referred to it in the menu of the wiki

**Comment by teichsta** [comment]
great, many thanks!

**Pull request merged**